hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/cache/EntityMapWarmer.java

```diff
@@ -0,0 +1,282 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.mono.cache;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.hedera.node.app.service.mono.context.properties.GlobalDynamicProperties;
+import com.hedera.node.app.service.mono.state.migration.AccountStorageAdapter;
+import com.hedera.node.app.service.mono.state.migration.TokenRelStorageAdapter;
+import com.hedera.node.app.service.mono.state.migration.UniqueTokenMapAdapter;
+import com.hedera.node.app.service.mono.state.virtual.EntityNumVirtualKey;
+import com.hedera.node.app.service.mono.state.virtual.UniqueTokenKey;
+import com.hedera.node.app.service.mono.state.virtual.entities.OnDiskAccount;
+import com.hedera.node.app.service.mono.utils.EntityNum;
+import com.hedera.node.app.service.mono.utils.EntityNumPair;
+import com.hedera.node.app.service.mono.utils.accessors.PlatformTxnAccessor;
+import com.hederahashgraph.api.proto.java.AccountAmount;
+import com.hederahashgraph.api.proto.java.AccountID;
+import com.hederahashgraph.api.proto.java.NftTransfer;
+import com.hederahashgraph.api.proto.java.TokenTransferList;
+import com.hederahashgraph.api.proto.java.TransactionBody;
+import com.hederahashgraph.api.proto.java.TransferList;
+import com.swirlds.common.system.Round;
+import com.swirlds.common.system.events.ConsensusEvent;
+import com.swirlds.virtualmap.VirtualMap;
+import java.util.Objects;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import org.bouncycastle.util.encoders.Hex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A proto-duction implementation of cache warming for {@link VirtualMap}, i.e. a quick-and-dirty
+ * implementation in order to get it out ASAP. The driving theory behind this class is that we
+ * can pre-fetch objects we access often in order to get better performance time.
+ *
+ * <p>This class is geared towards accounts, unique tokens (NFTs), and token relations, but could
+ * also be enhanced for other types.
+ */
+public class EntityMapWarmer {
+
+    private static final Logger log = LoggerFactory.getLogger(EntityMapWarmer.class);
+    private static EntityMapWarmer instance;
+    // Note: these suppliers need to be stored here as they are (instead of storing what
+    // the supplier references) in order for this class to get the latest copies of the
+    // underlying data structures
+    private final Supplier<AccountStorageAdapter> accountsStorageAdapter;
+    private final Supplier<UniqueTokenMapAdapter> nftsAdapter;
+    private final Supplier<TokenRelStorageAdapter> tokenRelsAdapter;
+    private final ThreadPoolExecutor threadpool;
```

Comment:
It might be worth discussing the direct use of ThreadPoolExecutor and the execute method.  Ideally, in my opinion, this class would use the ExecutorService interface, the submit method, and store the Future results in a list-per-round, so that instead of digging through a queue to find things when round warmup is cancelled, the list could be walked and any Future's that are not done cancelled (and the list removed).  That would work better with some of the other efforts in platform to improve management of threads and timing of critical events.

Reply:
Collecting and processing Futures of submitted tasks creates unwanted overhead with no benefit. The purpose of the warmer is to engage parallel disk I/O to populate underlying caches (`MerkleDb` read cache, OS file cache) without as little interference (synchronization) with transaction handling as possible. Cancelling pending warmup tasks is a defensive move preventing the system from overloading. There is no external  semantics for it and the `cancelPendingWarmups`  method should not even be public (see my comment above).

Reply:
I see.  It wasn't clear that we don't need to individually cancel the tasks, which changes the calculus.  It might be worth noting that it isn't necessary to create both Runnable and Future (one can use RunnableFuture to avoid the extra object), but if we can just clear the queue, then we don't need that.
I still suspect that we could get a lot better results overall using some of the more adaptable concepts, but using the ThreadPoolExecutor directly is reasonable in the short term.